examples/lox/src/language/type-system/lox-type-checking.ts

```diff
@@ -245,13 +244,15 @@ export class LoxTypeCreator extends AbstractLangiumTypeCreator {
                 },
                 // inference rule for accessing fields
                 inferenceRuleForFieldAccess: (domainElement: unknown) => isMemberCall(domainElement) && isFieldMember(domainElement.element?.ref) && domainElement.element!.ref.$container === node
-                    ? domainElement.element!.ref.name : 'N/A', // as an alternative, use 'InferenceRuleNotApplicable' instead, what should we recommend?
+                    ? domainElement.element!.ref.name : InferenceRuleNotApplicable,
             });
 
             // TODO conversion 'nil' to classes ('TopClass')!
             // any class !== all classes; here we want to say, that 'nil' is assignable to each concrete Class type!
             // this.typir.conversion.markAsConvertible(typeNil, this.classKind.getOrCreateTopClassType({}), 'IMPLICIT_EXPLICIT');
-            this.typir.conversion.markAsConvertible(this.primitiveKind.getPrimitiveType({ primitiveName: 'nil' })!, classType, 'IMPLICIT_EXPLICIT');
+            classType.addListener(type => {
+                this.typir.conversion.markAsConvertible(this.primitiveKind.getPrimitiveType({ primitiveName: 'nil' })!, type, 'IMPLICIT_EXPLICIT');
+            });
```

Comment:
I think this listener registration can be replaced by a Promise, because you are just resolving something. Then the code looks more synchronous using await/async.

Reply:
```ts
resolved(): Promise<Type> {
    return new Promise<Type>(listener => this.listeners.push(listener));
}
```

And then:

```ts
await classType.resolved();
this.typir.conversion. ...
```

But I am not entirely sure about this. Since you would await, your control flow might be blocked...

Reply:
> Since you would await, your control flow might be blocked...

Yes, and this is a problem when integrating Typir with Langium, since Typir hooks into the Langium process, which works in synchronous way, i.e. "there is no one who can awake you"